pyccel/utilities/metaclasses.py

```diff
@@ -0,0 +1,32 @@
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/master/LICENSE for full license details.     #
+#------------------------------------------------------------------------------------------#
+""" Module containing metaclasses which are useful for the rest of pyccel
+"""
+
+
+__all__ = (
+    'Singleton',
+    'ArgumentSingleton',
+)
+
+class Singleton(type):
+    """ Indicates that there is only one instance of the class
+    """
+    _instances = {}
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super().__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+class ArgumentSingleton(type):
+    """ Indicates that there is only one instance of the class
+    for any given set of arguments
+    """
+    _instances = {}
+    def __call__(cls, *args, **kwargs):
+        index = (cls, *args, *kwargs.values())
```

Comment:
I am worried that the order in which the keyword arguments are passed may play a role here. Further, I think that we should compare the keys as well in order to be safe. What about this?
```python
index = (cls, *args, *sorted(kwargs.items()))
```

Reply:
I can do this. I didn't think this class would be used for cases which had multiple arguments, only for simple things such as `LiteralTrue`

Reply:
Fixed in c4c7b0e

Reply:
Thanks